File: src-tauri/src/screen_mirror.rs
use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};
use std::collections::HashMap;

/// 投屏质量设置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenMirrorQuality {
    pub resolution: String,
    pub bitrate: u32,
    pub framerate: u32,
    pub codec: String,
}

/// 投屏配置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenMirrorConfig {
    pub quality: ScreenMirrorQuality,
    #[serde(rename = "showTouches")]
    pub show_touches: bool,
    #[serde(rename = "stayAwake")]
    pub stay_awake: bool,
    #[serde(rename = "turnScreenOff")]
    pub turn_screen_off: bool,
    #[serde(rename = "recordScreen")]
    pub record_screen: bool,
    #[serde(rename = "audioEnabled")]
    pub audio_enabled: bool,
    #[serde(rename = "controlEnabled")]
    pub control_enabled: bool,
}

/// 投屏状态
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ScreenMirrorStatus {
    Disconnected,
    Connecting,
    Connected,
    Streaming,
    Paused,
    Error,
}

/// 投屏会话信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenMirrorSession {
    pub id: String,
    pub device_serial: String,
    pub device_name: Option<String>,
    pub status: ScreenMirrorStatus,
    pub config: ScreenMirrorConfig,
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub error: Option<String>,
    pub process_id: Option<u32>,
    pub server_port: Option<u16>,
    pub video_socket: Option<u16>,
    pub control_socket: Option<u16>,
}

/// 投屏统计信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenMirrorStats {
    pub session_id: String,
    pub duration: u64,
    pub frames_received: u64,
    pub frames_dropped: u64,
    pub average_fps: f32,
    pub current_bitrate: f32,
    pub network_latency: u32,
}

/// 投屏控制事件类型
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
pub enum ScreenMirrorControlEvent {
    Touch(TouchEventData),
    Key(KeyEventData),
    Scroll(ScrollEventData),
}

/// 触摸事件数据
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TouchEventData {
    pub action: String, // "down", "up", "move"
    pub x: f32,
    pub y: f32,
    pub pointer_id: Option<u32>,
}

/// 按键事件数据
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KeyEventData {
    pub action: String, // "down", "up"
    pub key_code: u32,
    pub meta_state: Option<u32>,
}

/// 滚动事件数据
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScrollEventData {
    pub x: f32,
    pub y: f32,
    pub delta_x: f32,
    pub delta_y: f32,
}

/// 投屏设备信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenMirrorDevice {
    pub serial: String,
    pub name: Option<String>,
    pub model: Option<String>,
    pub resolution: Option<String>,
    pub density: Option<u32>,
    pub orientation: Option<String>,
    #[serde(rename = "isSupported")]
    pub is_supported: bool,
    #[serde(rename = "supportedCodecs")]
    pub supported_codecs: Vec<String>,
}

/// 投屏错误信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenMirrorError {
    pub code: String,
    pub message: String,
    pub details: Option<String>,
    pub timestamp: DateTime<Utc>,
    pub session_id: Option<String>,
}

impl Default for ScreenMirrorQuality {
    fn default() -> Self {
        Self {
            resolution: "1280x720".to_string(),
            bitrate: 2,
            framerate: 30,
            codec: "h264".to_string(),
        }
    }
}

impl Default for ScreenMirrorConfig {
    fn default() -> Self {
        Self {
            quality: ScreenMirrorQuality::default(),
            show_touches: false,
            stay_awake: true,
            turn_screen_off: false,
            record_screen: false,
            audio_enabled: true,
            control_enabled: true,
        }
    }
}

impl ScreenMirrorSession {
    pub fn new(device_serial: String, config: ScreenMirrorConfig) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            device_serial,
            device_name: None,
            status: ScreenMirrorStatus::Disconnected,
            config,
            start_time: None,
            end_time: None,
            error: None,
            process_id: None,
            server_port: None,
            video_socket: None,
            control_socket: None,
        }
    }

    pub fn start(&mut self) {
        self.status = ScreenMirrorStatus::Connecting;
        self.start_time = Some(Utc::now());
        self.error = None;
    }

    pub fn set_connected(&mut self, process_id: u32, server_port: u16) {
        self.status = ScreenMirrorStatus::Connected;
        self.process_id = Some(process_id);
        self.server_port = Some(server_port);
    }

    pub fn set_streaming(&mut self) {
        self.status = ScreenMirrorStatus::Streaming;
    }

    pub fn set_error(&mut self, error: String) {
        self.status = ScreenMirrorStatus::Error;
        self.error = Some(error);
        self.end_time = Some(Utc::now());
    }

    pub fn stop(&mut self) {
        self.status = ScreenMirrorStatus::Disconnected;
        self.end_time = Some(Utc::now());
        self.process_id = None;
        self.server_port = None;
        self.video_socket = None;
        self.control_socket = None;
    }
}

/// 预设的投屏质量配置
pub fn get_quality_presets() -> HashMap<String, ScreenMirrorQuality> {
    let mut presets = HashMap::new();
    
    presets.insert("ultra".to_string(), ScreenMirrorQuality {
        resolution: "auto".to_string(),
        bitrate: 8,
        framerate: 60,
        codec: "h264".to_string(),
    });
    
    presets.insert("high".to_string(), ScreenMirrorQuality {
        resolution: "1920x1080".to_string(),
        bitrate: 4,
        framerate: 30,
        codec: "h264".to_string(),
    });
    
    presets.insert("medium".to_string(), ScreenMirrorQuality {
        resolution: "1280x720".to_string(),
        bitrate: 2,
        framerate: 30,
        codec: "h264".to_string(),
    });
    
    presets.insert("low".to_string(), ScreenMirrorQuality {
        resolution: "854x480".to_string(),
        bitrate: 1,
        framerate: 15,
        codec: "h264".to_string(),
    });
    
    presets
}

File: src/types/screenMirror.ts
// 投屏质量设置
export interface ScreenMirrorQuality {
  resolution: string; // 分辨率，如 "1920x1080", "1280x720", "auto"
  bitrate: number; // 比特率，单位 Mbps
  framerate: number; // 帧率，单位 fps
  codec: "h264" | "h265"; // 编码格式
}

// 投屏配置
export interface ScreenMirrorConfig {
  quality: ScreenMirrorQuality;
  showTouches: boolean; // 显示触摸点
  stayAwake: boolean; // 保持屏幕常亮
  turnScreenOff: boolean; // 投屏时关闭设备屏幕
  powerOffOnClose: boolean; // 关闭时关机
  noPowerOn: boolean; // 启动时不开机
  recordScreen: boolean; // 是否录制屏幕
  audioEnabled: boolean; // 是否传输音频
  controlEnabled: boolean; // 是否启用控制功能
  alwaysOnTop: boolean; // 投屏窗口始终置顶
  fullscreen: boolean; // 全屏模式
  windowTitle: string; // 窗口标题
  crop: string; // 屏幕裁剪，格式为"width:height:x:y"
  maxFps: number; // 最大帧率限制
  lockVideoOrientation: "unlocked" | "initial" | "portrait" | "landscape"; // 锁定视频方向
  rotation: number; // 屏幕旋转角度 (0, 1, 2, 3)
}

// 投屏状态
export type ScreenMirrorStatus = 
  | "disconnected" // 未连接
  | "connecting" // 连接中
  | "connected" // 已连接
  | "streaming" // 投屏中
  | "paused" // 暂停
  | "error"; // 错误状态

// 投屏会话信息
export interface ScreenMirrorSession {
  id: string;
  deviceSerial: string;
  deviceName?: string;
  status: ScreenMirrorStatus;
  config: ScreenMirrorConfig;
  startTime?: Date;
  endTime?: Date;
  error?: string;
  processId?: number; // scrcpy 进程ID
  serverPort?: number; // 服务器端口
  videoSocket?: number; // 视频流端口
  controlSocket?: number; // 控制端口
}

// 投屏统计信息
export interface ScreenMirrorStats {
  sessionId: string;
  duration: number; // 投屏时长（秒）
  framesReceived: number; // 接收帧数
  framesDropped: number; // 丢帧数
  averageFps: number; // 平均帧率
  currentBitrate: number; // 当前比特率
  networkLatency: number; // 网络延迟（毫秒）
}

// 投屏控制事件
export interface ScreenMirrorControlEvent {
  type: "touch" | "key" | "scroll" | "gesture";
  timestamp: number;
  data: any; // 具体的控制数据
}

// 触摸事件数据
export interface TouchEventData {
  action: "down" | "up" | "move";
  x: number;
  y: number;
  pointerId?: number;
}

// 按键事件数据
export interface KeyEventData {
  action: "down" | "up";
  keyCode: number;
  metaState?: number;
}

// 滚动事件数据
export interface ScrollEventData {
  x: number;
  y: number;
  deltaX: number;
  deltaY: number;
}

// 投屏设备信息
export interface ScreenMirrorDevice {
  serial: string;
  name?: string;
  model?: string;
  resolution?: string;
  density?: number;
  orientation?: "portrait" | "landscape";
  isSupported: boolean; // 是否支持投屏
  supportedCodecs: string[]; // 支持的编码格式
}

// 投屏错误类型
export interface ScreenMirrorError {
  code: string;
  message: string;
  details?: string;
  timestamp: Date;
  sessionId?: string;
}

// 预设的投屏质量配置
export const SCREEN_MIRROR_QUALITY_PRESETS: Record<string, ScreenMirrorQuality> = {
  "ultra": {
    resolution: "auto",
    bitrate: 12,
    framerate: 60,
    codec: "h264"
  },
  "high": {
    resolution: "1920x1080",
    bitrate: 8,
    framerate: 60,
    codec: "h264"
  },
  "balanced": {
    resolution: "1920x1080",
    bitrate: 4,
    framerate: 30,
    codec: "h264"
  },
  "medium": {
    resolution: "1280x720",
    bitrate: 2,
    framerate: 30,
    codec: "h264"
  },
  "low": {
    resolution: "854x480",
    bitrate: 1,
    framerate: 15,
    codec: "h264"
  },
  "gaming": {
    resolution: "1920x1080",
    bitrate: 10,
    framerate: 60,
    codec: "h264"
  },
  "powerSaving": {
    resolution: "1280x720",
    bitrate: 1,
    framerate: 15,
    codec: "h264"
  },
  "h265High": {
    resolution: "1920x1080",
    bitrate: 6,
    framerate: 60,
    codec: "h265"
  },
  "h265Balanced": {
    resolution: "1920x1080",
    bitrate: 3,
    framerate: 30,
    codec: "h265"
  }
};

// 默认投屏配置
export const DEFAULT_SCREEN_MIRROR_CONFIG: ScreenMirrorConfig = {
  quality: SCREEN_MIRROR_QUALITY_PRESETS.high,
  showTouches: false,
  stayAwake: true,
  turnScreenOff: false,
  powerOffOnClose: false,
  noPowerOn: false,
  recordScreen: false,
  audioEnabled: true,
  controlEnabled: true,
  alwaysOnTop: false,
  fullscreen: false,
  windowTitle: "HOUT投屏",
  crop: "",
  maxFps: 60,
  lockVideoOrientation: "unlocked",
  rotation: 0
};
